storage/src/main/java/io/delta/storage/internal/ThreadUtils.java

```diff
@@ -0,0 +1,117 @@
+        //since T exception or T result cannot be used in static method
+        List<Exception> exceptionHolder = new ArrayList<>(1);
+        List<T> resultHolder = new ArrayList<>(1);
+        Thread thread = new Thread(threadName) {
+            @Override
+            public void run() {
+                try {
+                    resultHolder.add(body.call());
+                } catch (Exception ex) {
+                    exceptionHolder.add(ex);
+                }
+            }
+        };
+        thread.setDaemon(isDaemon);
+        thread.start();
+        try {
+            thread.join();
+        } catch (InterruptedException e) {
+            throw new InterruptedIOException(e.getMessage());
+        }
+        if (!exceptionHolder.isEmpty()) {
+            Exception realException = exceptionHolder.get(0);
+            // Remove the part of the stack that shows method calls into this helper method
+            // This means drop everything from the top until the stack element
+            // ThreadUtils.runInNewThread(), and then drop that as well (hence the + 1 to start index).
+            List<StackTraceElement> currentThreadStackTrace =
+                    Arrays.asList(Thread.currentThread().getStackTrace());
+            int startIndex = currentThreadStackTrace.indexOf(ThreadUtils.class.getSimpleName()) + 1;
+            int endIndex = currentThreadStackTrace.size();
+            List<StackTraceElement> baseStackTrace;
+            if (startIndex <= 0 || startIndex >= endIndex) {
+                baseStackTrace = Collections.emptyList();
+            } else {
+                baseStackTrace =
+                        currentThreadStackTrace.subList(startIndex, currentThreadStackTrace.size());
+            }
+            // Remove the part of the new thread stack that shows methods call from this helper method
+            // This means take everything from the top until the stack element
+            StackTraceElement[] realExceptionStackTrace = realException.getStackTrace();
+            List<StackTraceElement> extraStackTrace = new ArrayList<>();
+            for (StackTraceElement st : realExceptionStackTrace) {
+                if (st.getClassName().contains(ThreadUtils.class.getSimpleName())) {
+                    break;
+                }
+                extraStackTrace.add(st);
+            }
+            // Combine the two stack traces, with a place holder just specifying that there
+            // was a helper method used, without any further details of the helper
+            List<StackTraceElement> placeHolderStackElem = Arrays.asList(
+                    new StackTraceElement(
+                            String.format(
+                                    "... run in separate thread using $s",
+                                    ThreadUtils.class.getSimpleName(),
+                                    " static method runInNewThread"),
+                            "",
```

Comment:
inline comments beside each of these params telling us what they are for. as a reader, I don't know what -1 is for